### PR TITLE
add max leeway configuration and validation

### PR DESCRIPTION
**Pull request by kodjo-anipah**
part of Graylog2/graylog-plugin-cloud#1149
Restrict configuration of Time Size Optimising Retention on cloud.
## Description

- Added new config `time_size_optimizing_retention_max_leeway` to restrict min and max lifetime leeway
- cloud only add migration for default index set configuration to be set to time size optimizing
- on cloud, hide max latch and automatically set max to min + max leeway. 
- hint the user about what the calculated resulted max value would be 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (non-breaking change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: graylog2-server/src/main/java/org/graylog2/configuration/ElasticsearchConfiguration.java [hunk 1]
```diff
@@ -132,6 +134,9 @@ public class ElasticsearchConfiguration {
     @Parameter(value = TIME_SIZE_OPTIMIZING_RETENTION_MAX_LIFETIME)
     private Period timeSizeOptimizingRotationMaxLifeTime = TimeBasedSizeOptimizingStrategyConfig.DEFAULT_LIFETIME_MAX;
 
+    @Parameter(value = TIME_SIZE_OPTIMIZING_RETENTION_FIXED_LEEWAY)
+    private Period timeSizeOptimizingRotationFixedLeeway = Period.days(1);
```
Comment (mpfz0r) [inline graylog2-server/src/main/java/org/graylog2/configuration/ElasticsearchConfiguration.java:138]: for on-prem we don't want this to be on by default.
can we set the default to null? and only validate it if it's set?

File: graylog2-server/src/main/java/org/graylog2/indexer/IndexSetValidator.java [hunk 2]
```diff
@@ -120,9 +120,10 @@ public Violation validateRotation(RotationStrategyConfig rotationStrategyConfig)
                 return Violation.create(f("%s <%s> is shorter than %s <%s>", INDEX_LIFETIME_MAX, config.indexLifetimeMax(),
                         INDEX_LIFETIME_MIN, config.indexLifetimeMin()));
             }
-            if (leeway.toStandardSeconds().isLessThan(elasticsearchConfiguration.getTimeSizeOptimizingRotationPeriod().toStandardSeconds())) {
+
+            if (leeway.toStandardSeconds().isLessThan(elasticsearchConfiguration.getTimeSizeOptimizingRotationFixedLeeway().toStandardSeconds())) {
```
Comment (mpfz0r) [inline graylog2-server/src/main/java/org/graylog2/indexer/IndexSetValidator.java:124]: i would keep this validation and add a new separate fixed leeway validation (which is off by default (see above comment))

**Review by mpfz0r — CHANGES_REQUESTED** [review]
two comments, rest looks good on a first glance

File: graylog2-server/src/main/java/org/graylog2/configuration/ElasticsearchConfiguration.java [hunk 3]
```diff
@@ -264,6 +269,10 @@ public Period getTimeSizeOptimizingRotationMaxLifeTime() {
         return timeSizeOptimizingRotationMaxLifeTime;
     }
 
+    public Period getTimeSizeOptimizingRotationFixedLeeway() {
```
Comment (mpfz0r) [inline graylog2-server/src/main/java/org/graylog2/configuration/ElasticsearchConfiguration.java:272]:
```suggestion
    @Nullable
    public Period getTimeSizeOptimizingRotationFixedLeeway() {
```

File: graylog2-server/src/test/java/org/graylog2/indexer/IndexSetValidatorTest.java [hunk 4]
```diff
@@ -169,6 +169,8 @@ public void validateMaxRetentionPeriod() {
                 assertThat(v.message()).contains("effective index retention period of P1W3D")
         );
 
+        when(elasticsearchConfiguration.getTimeSizeOptimizingRotationFixedLeeway()).thenReturn(Period.days(28));
+
```
Comment (mpfz0r) [inline graylog2-server/src/test/java/org/graylog2/indexer/IndexSetValidatorTest.java:173]:
```suggestion
```

we don't need this anymore

File: graylog2-server/src/test/java/org/graylog2/indexer/IndexSetValidatorTest.java [hunk 5]
```diff
@@ -194,6 +196,7 @@ public void validateMaxRetentionPeriod() {
     @Test
     public void timeBasedSizeOptimizingOnlyWithMultipleOfDays() {
         when(elasticsearchConfiguration.getTimeSizeOptimizingRotationPeriod()).thenReturn(Period.days(1));
+        when(elasticsearchConfiguration.getTimeSizeOptimizingRotationFixedLeeway()).thenReturn(Period.days(1));
```
Comment (mpfz0r) [inline graylog2-server/src/test/java/org/graylog2/indexer/IndexSetValidatorTest.java:199]:
```suggestion
```
same here

**Review by mpfz0r — APPROVED** [review]
LGTM :+1:

**Review by kodjo-anipah — COMMENTED** [review]
image
looks good now

**Pull request merged**